### DISCUSSION TITLE
ljspeech\glow_tts - Character '͡' not found in the vocabulary. Discarding it.

### DISCUSSION THREAD

**Discussion by koayst** [coqui-ai/TTS#1842] [General Q&A]
Hi,

I am training the model under "recipes\ljspeech\glow_tts\train_glowstts.py" (using CPU) and have also downloaded the LJSpeech-1.1 dataset.  Below shows the output of the training.

 > Setting up Audio Processor...
 | > sample_rate:22050
 | > resample:False
 | > num_mels:80
 | > log_func:np.log10
 | > min_level_db:-100
 | > frame_shift_ms:None
 | > frame_length_ms:None
 | > ref_level_db:20
 | > fft_size:1024
 | > power:1.5
 | > preemphasis:0.0
 | > griffin_lim_iters:60
 | > signal_norm:True
 | > symmetric_norm:True
 | > mel_fmin:0
 | > mel_fmax:None
 | > pitch_fmin:0.0
 | > pitch_fmax:640.0
 | > spec_gain:20.0
 | > stft_pad_mode:reflect
 | > max_norm:4.0
 | > clip_norm:True
 | > do_trim_silence:True
 | > trim_db:45
 | > do_sound_norm:False
 | > do_amp_to_db_linear:True
 | > do_amp_to_db_mel:True
 | > do_rms_norm:False
 | > db_level:None
 | > stats_path:None
 | > base:10
 | > hop_length:256
 | > win_length:1024
 | > Found 13100 files in C:\Users\koay_seng_tian\Downloads\TTS\recipes\ljspeech\LJSpeech-1.1
 > Using CUDA: False
 > Number of GPUs: 0
 > Model has 28610257 parameters
 > EPOCH: 0/1000
 --> C:\Users\koay_seng_tian\Downloads\TTS\recipes\ljspeech\glow_tts\run-August-08-2022_08+31PM-903a77c1
[*] Pre-computing phonemes...
  0%|          | 3/12969 [00:01<2:09:25,  1.67it/s]ɪnstɛd əv weɪtɪŋ ðɛɹ, ɔzwɔld əpɛɹəntli wɛnt æz fɑɹ əweɪ æz hi kʊd ænd bɔɹdɪd ðə fɚst oʊk klɪf bʌs wɪt͡ʃ keɪm əlɔŋ
 [!] Character '͡' not found in the vocabulary. Discarding it.

I wonder why there is this "Character not found in the vocabulary.  Discarding it" message.
Is it some error I need to be concerned about?  Or some setup I didn't do properly hence the error ?

I did a trace of the error and at gruut_wrapper.py, there is a function called "phonemize_gruut" in which one of the parameters "tie" is set to False.  It also mentioned that:

tie (bool, optional) : When True use a '͡' character between consecutive characters of a single phoneme. Else separate phoneme                with '_'. This option requires espeak>=1.49. Default to False.   

The statement seems to say that the "error" message on character not found in the vocabulary is a problem.

What should I do ?

**Comment by harmlessman** [comment]
The reason for such warnings is that there is  '͡' in metadata.csv in ljspeech, but  '͡' is not declared in characters.
If you want to remove the warning, add  '͡' to the characters or punctuations  in your characters_class.

But if  '͡' is not used a lot and only one or two sentences are in it, it doesn't matter much to ignore it.

I also saw a lot of such warnings when I was learning with my dataset, but if there were fewer sentences with symbols (words) that generated a particular warning, ignoring them did not affect the results.

**Reply by koayst** [reply]
Hi @harmlessman,

I did a trace and for training the "train_glowtts.py", it leads me to TTS\tts\utils\text\characters.py.  Am I correct to look for  "characters_class" here?

![image](https://user-images.githubusercontent.com/59504709/183623199-152aff2e-f261-44a8-8f9a-3ce81fd33897.png)

**Reply by koayst** [reply]
Hi,

See the screen shot I took just now.  The purple highligted English text is the text to be translated to phoneme.  It seems the character "  '͡'  " is not correctly depicting the problem.   From the screen shot, it looks like an arc connecting two characters.  I am still new to IPA (International Phonetic Alphabet) and according to what I understand from [stackexchange](https://linguistics.stackexchange.com/questions/42956/t%CD%A1%CA%83-vs-%CA%A7-vs-t%CA%83), "tie bar" is a legit character for phoneme.
 
![image](https://user-images.githubusercontent.com/59504709/183626677-7cc285b5-2586-44af-8cf2-bac5cc1e93d6.png)

In gruut_wrapper.py, there is a function called "phonemize_gruut" in which one of the parameters "tie" is set to False.  So I was wondering why this tie bar character was generated even thought "tie" is set to False.

**Comment by harmlessman** [comment]
1. What I said is `character` or `punctuations` in the config part like this picture

![image](https://user-images.githubusercontent.com/87223285/183785265-5fd1a2de-2537-4fe9-870a-fb93251c29a0.png)
But if you use IPA, you can add a symbol where you say it.
I don't know exactly what A is used for, but if it's character (ex - 'a', 'b'...), it can be added to `character`, and if it's symbol (','', '?'), it can be added to `punctuations`.


2.why   ͡   look like  '͡'  ? 
because  '͡'  is `'  +   ͡   + '`
 
In order not to raise a warning, you need to add  ͡  to the `character` or `punctuations` 
 `add ͡`


3.
![image](https://user-images.githubusercontent.com/87223285/183787723-05ccbff4-8de3-4134-9732-20dd441dfa02.png)

I can't say for sure because it's not the source code I made, In `phonemize_gruut.py`, `tie` argument is not used.
That is why` ͡ ` is appear.
